service.go

```diff
@@ -37,40 +37,143 @@ type Service struct {
 	sighupHandler func()
 }
 
-// New creates a new named service and allows for customization through functional options.
-func New(name, version string, oo ...OptionFunc) (*Service, error) {
+const fieldSetMsg = "Setting service property '%v' with '%v'"
+
+// Builder gathers all required properties to
+// construct a Patron service.
+type Builder struct {
+	errors        []error
+	name          string
+	version       string
+	cps           []Component
+	routes        []http.Route
+	middlewares   []http.MiddlewareFunc
+	acf           http.AliveCheckFunc
+	rcf           http.ReadyCheckFunc
+	termSig       chan os.Signal
+	sighupHandler func()
+}
+
+// NewBuilder initiates the Service builder chain.
+// The builder contains default values for Alive/Ready checks,
+// the SIGHUP handler and its version.
+func NewBuilder(name, version string) *Builder {
+	var errs []error
+
 	if name == "" {
-		return nil, errors.New("name is required")
+		errs = append(errs, errors.New("name is required"))
 	}
-
 	if version == "" {
 		version = "dev"
 	}
 
-	s := Service{
-		cps:           []Component{},
+	return &Builder{
+		errors:        errs,
+		name:          name,
+		version:       version,
 		acf:           http.DefaultAliveCheck,
 		rcf:           http.DefaultReadyCheck,
 		termSig:       make(chan os.Signal, 1),
 		sighupHandler: func() { log.Info("SIGHUP received: nothing setup") },
-		middlewares:   []http.MiddlewareFunc{},
 	}
+}
 
-	err := Setup(name, version)
-	if err != nil {
-		return nil, err
+// WithRoutes adds routes to the default HTTP component.
+func (b *Builder) WithRoutes(rr []http.Route) *Builder {
```

Comment:
as we are aggregating the errors, we can always check if there are errors, an d potentially "do nothing" as it will anyway return an error in the end. e.g. `if len(b.errors) > 0; return b;`

Reply:
Yes we could and we have in the past, but since we are checking in the build method for previous errors we can avoid this making the code smaller and much easier to test. In the end the error will be returned by the build method...